src/adc.rs

```diff
@@ -0,0 +1,268 @@
+
+use crate::stm32;
+
+use crate::gpio::*;
+
+/// Analog to Digital converter interface
+pub struct Adc {
+    rb: stm32::ADC,
+    sample_time: AdcSampleTime,
+}
+
+/// ADC Sampling rate
+///
+/// Options for the sampling time, each is T + 0.5 ADC clock cycles.
+/// So a setting of T_55 results in a sample time of 55.5 ADC clock cycles.
+pub enum AdcSampleTime {
+    T_1,
+    T_7,
+    T_13,
+    T_28,
+    T_41,
+    T_55,
+    T_71,
+    T_239,
+}
+
+impl AdcSampleTime {
+    fn write_bits(&self, adc: &mut stm32::ADC) {
+        unsafe {
+            adc.smpr.write(|w| w.smpr().bits(
+                match self {
+                    AdcSampleTime::T_1   => 0b000_u8,
+                    AdcSampleTime::T_7   => 0b001_u8,
+                    AdcSampleTime::T_13  => 0b010_u8,
+                    AdcSampleTime::T_28  => 0b011_u8,
+                    AdcSampleTime::T_41  => 0b100_u8,
+                    AdcSampleTime::T_55  => 0b101_u8,
+                    AdcSampleTime::T_71  => 0b110_u8,
+                    AdcSampleTime::T_239 => 0b111_u8,
+                }
+            ));
+        }
+    }
+
+    /// Get the default sample time (currently 239.5 cycles)
+    pub fn default() -> Self {
+        AdcSampleTime::T_239
+    }
+}
+
+macro_rules! adc_pins {
+    ($($pin:ty => $chan:expr),+ $(,)*) => {
+        $(
+            impl Channel<Adc> for $pin {
+                type ID = u8;
+
+                fn channel() -> u8 { $chan }
+            }
+        )+
+    };
+}
+
+#[cfg(any(feature = "stm32f042", feature = "stm32f030", feature = "stm32f070",))]
+adc_pins!(
+    gpioa::PA0<Analog> => 0_u8,
+    gpioa::PA1<Analog> => 1_u8,
+    gpioa::PA2<Analog> => 2_u8,
+    gpioa::PA3<Analog> => 3_u8,
+    gpioa::PA4<Analog> => 4_u8,
+    gpioa::PA5<Analog> => 5_u8,
+    gpioa::PA6<Analog> => 6_u8,
+    gpioa::PA7<Analog> => 7_u8,
+    gpiob::PB0<Analog> => 8_u8,
+    gpiob::PB1<Analog> => 9_u8,
+);
+
+#[cfg(any(feature = "stm32f030", feature = "stm32f070",))]
+adc_pins!(
+    gpioc::PC0<Analog> => 10_u8,
+    gpioc::PC1<Analog> => 11_u8,
+    gpioc::PC2<Analog> => 12_u8,
+    gpioc::PC3<Analog> => 13_u8,
+    gpioc::PC4<Analog> => 14_u8,
+    gpioc::PC5<Analog> => 15_u8,
+);
+
+/// Internal temperature sensor (ADC Channel 16)
+pub struct VTemp;
+
+/// Internal voltage reference (ADC Channel 17)
+pub struct VRef;
+
+/// Battery reference voltage (ADC Channel 18)
+pub struct VBat;
+
+adc_pins!(
+    VTemp => 16_u8,
+    VRef  => 17_u8,
+    VBat  => 18_u8,
+);
+
+impl VTemp {
+    pub fn new() -> Self { VTemp {} }
+
+    /// Enable the internal temperature sense, this has a wake up time
+    /// t<sub>START</sub> which can be found in your micro's datasheet, you
+    /// must wait at least that long after enabling before taking a reading.
+    /// Remember to disable when not in use.
+    pub fn enable(&mut self) {
+        let adc = unsafe{ &*stm32::ADC::ptr() };
+        adc.ccr.write(|w| w.tsen().set_bit());
```

Comment:
Why prohibit leaving multiple on at a time? (of course, then you'll have to consider race conditions)

Reply:
Ah, I have misunderstood how to operate the registers slightly, will correct. Most of my uses of `write` should instead by `modify`.